Create IteratorBuilder to expose potential bugs

milleruntime opened the pull request:
* Create builder classes for iterator stack creation
* Replace IterLoad class with new IteratorBuilder and refactor classes
that use it to determine whether or not there is a bug in iterator code

keith-turner commented:
> Replace IterLoad class with new IteratorBuilder and refactor classes that use it to determine whether or not there is a bug in iterator code

@milleruntime  how do these changes help find a bug?

milleruntime commented:
> > Replace IterLoad class with new IteratorBuilder and refactor classes that use it to determine whether or not there is a bug in iterator code
> 
> @milleruntime how do these changes help find a bug?

I was working on these changes when I thought there was a bug (see #2601). But it turns out that it is not a bug so I can removed the validation check I added to the `IteratorBuilder`. This check was throwing an error is some of the tests.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/conf/IteratorBuilder.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Pass true to useAccumuloClassLoader or false otherwise. If true, a context is required.
+     */
+    IteratorBuilderOptions useClassLoader(boolean useAccumuloClassLoader);
```
Instead of passing in a true/false, could supply the context here. Supplying a context implies true.

```suggestion
    IteratorBuilderOptions useClassLoaderContext(String context);
```

And, class loading always uses a class loader, so it's a bit confusing if the value is false.

milleruntime replied:
Ah yeah, I missed reverting this, thanks.

ctubbsii reviewed (commented):
o.a.a.core.conf doesn't seem like the right location for this. That is for AccumuloConfiguration related things. This is more an internal utility for instantiating iterators for scanning. Could be moved to an existing iterator impl package if there is one.

milleruntime commented:
I moved the classes into the core iteratorsImpl package in my latest commits. I also renamed `IterConfigUtil` to `IteratorConfigUtil` and moved it as well. This change was basically to finish up the work I did previously with cleaning up iterator code.

milleruntime commented:
@ctubbsii @keith-turner I think this is ready to go now.

ctubbsii commented:
> @ctubbsii @keith-turner I think this is ready to go now.

Okay, thanks. I'll try to take a look asap.

milleruntime commented:
Anyone get a chance to peek at this?

milleruntime reviewed (commented):
I am fairly confident that these changes are just cleaning up some of the internal code. If anyone has any more feedback, I can do a follow on task.
